File: exercism/python/bob/bob.py
#
# Skeleton file for the Python "Bob" exercise.
#
import re


def hey(foo):
    bar = foo.strip()

    if not len(bar):
        return 'Fine. Be that way!'
    elif re.search('[a-z0-9]+.*\?$', bar):
        return 'Sure.'
    elif all_upper(bar):
        return 'Whoa, chill out!'
    return 'Whatever.'


def all_upper(check_str):
    return not re.search('[a-z]', check_str) and re.search('[A-Z]+', check_str) \
        and check_str == check_str.upper()

File: exercism/haskell/bob/src/Bob.hs
module Bob
  ( responseFor
  ) where

import           Data.Char as C
import           Data.Text as T

responseFor :: String -> String
responseFor xs = responseFor' . T.strip $ T.pack xs

responseFor' :: Text -> String
responseFor' ys
  | ys == T.pack "" = "Fine. Be that way!"
  | shouting ys = "Whoa, chill out!"
  | T.last ys == '?' = "Sure."
  | otherwise = "Whatever."

shouting :: Text -> Bool
shouting t
  | T.last t == '!' = shouting' t
  | T.all (not . C.isAlpha) t = False
  | otherwise = shouting' t

shouting' :: Text -> Bool
shouting' = T.all (not . C.isLower)
